### ISSUE TITLE
React data model does not update when moving rows

### ISSUE THREAD

**Issue by henricook** [handsontable/handsontable#7517]
### Description
When you are using handsontable-react and bind a `data` model to your HOT instance, it generally updates whenever changes are made (e.g. with setDataAtCell).

It doesn't update for row moves however, as proven by the fiddle below. This took me hours to find, now I have to call setData(instance.getData()) from various places to make sure the model stays in sync.

### Steps to reproduce
1. See fiddle for small example

### Demo
https://jsfiddle.net/uhLx59bt/

### Your environment
* Handsontable version: 8.3.0
* Browser Name and version: Firefox 85
* Operating System: Linux

**Comment by AMBudnik** [comment]
Hi @henricook 

A similar topic has been raised here https://github.com/handsontable/react-handsontable/issues/137 and as you correctly spotted `getData()` is the method to use when you want to get the current states of the dataset.

**Comment by henricook** [comment]
Thanks for tying that in, i wouldn't call moving a row a sorting operation though. It's very much a permanent operation and not something I'd expect to be reversed later. Is there another way of moving rows?

**Comment by AMBudnik** [comment]
Yes, there are 3 ways to move rows.

![image](https://user-images.githubusercontent.com/10757813/107219042-cdd15080-6a10-11eb-8041-b506e31f0cc8.png)

1. use `manualRowMove: true` and allow users to move them as they want but do not change the initial order of rows
2. use the initial order of rows, like `manualRowMove: [1, 4]`
3. turn the plugin in the initialization settings and do not change the order of rows, but use the [`moveRow`](https://handsontable.com/docs/8.3.0/ManualRowMove.html#moveRow)/ [`moveRows`](https://handsontable.com/docs/8.3.0/ManualRowMove.html#moveRows) method

**Comment by henricook** [comment]
Thanks, i'm doing (1) and (3) in my application. So are you saying that moving rows is always perceived as _sorting_ so therefore the react data model will never be updated (but getData() is)?

**Comment by AMBudnik** [comment]
Please dont get me wrong. I am not saying that sorting and moving have some permanent or temporary effect. But there is some truth in that sentence.

We can see here https://jsfiddle.net/handsoncode/62qktLf9/ that we get the same result for state and `getData()` when we use 'remove row' from the context menu. But we get different data when we use 'hide row'.

**Comment by henricook** [comment]
Thanks, I was reading back your statements to ensure I understand. I find it's useful for clear outcomes :-)

If sorting is temporary, and moving is permanent, I would say that moving should be represented in the react data model, or we should have a way of saying "this move is a permanent move" - what do you think?

**Comment by AMBudnik** [comment]
> "this move is a permanent move" - what do you think

In Handsontable as long as you don't block the moving for users via `beforeRowMove` each move can be temporary. Meaning, if a user decides to move the row he can do it.

But I fully agree that there's a huge space for improvement when it comes to our state management definition in the documentation. We should not avoid creating more tutorials for such an important feature of React.

**Comment by henricook** [comment]
Sure, my assumption was that the react data model would always === instance.getData()

**Comment by wojciechczerniak** [comment]
It's not about change permanency. 

First, `sourceData` is not altered internally and there is nothing to sync, we're storing sorted or moved indexes as a meta data. That's how Handsontable handles this internally for performance reasons.

Second, it's against React (and other framework) design best practices. Components should not modify the `store` by reference. All changes from the child components should be announced by events and synced to the `store` with mutation actions. It's also recommended to break the reference to `data` property when passed to Handsontable to avoid data mutation inside the cells. We're not enforcing this because there are different needs and the dev can decide how to handle data exchange with his app

For performance reasons mentioned, I wouldn't recommend syncing it in every use case scenario. This may work well only for small data sets.

TL;DR; This issue is correctly marked as "Documentation", we should describe this better and link to React/Redux docs.

**Comment by henricook** [comment]
For my particular use case every time a move happens i'm saving `instance.getData()` to the server - and as of this not-bug i'm also calling setData(instance.getData()) in my react app. Thanks for pointing out that this is likely only a good idea because I'm dealing with small datasets

**Comment by henricook** [comment]
@wojciechczerniak Thanks for your comments, thinking more about them now. I was using setData(instance.getData()) as a workaround to 'make row moves permanent' but this has the unfortunate result of destroying my undo/redo buffer. Which means I have to choose between permanent row moves, and undo/redo 😂  - a tough call

I don't think breaking the reference to data (i.e. removing `data={data}` from my `<HotTable` in react terms) is a big problem for me, so I can do that, but then instead of doing `setData(instance.getData())` I'd have to do `instance.loadData(instance.getData())`.. which I think will still destroy my undo/redo buffer

Is there a way to replace the data in the table, without all the reinitialisation and destruction that comes with loadData? 

I hope I've explained this clearly enough but please ask if not, I want row moves to be considered updates to the data and not be semi permanent

Usage example: I'm sorting sections of the table automatically, so a user can enter a new row with a date and all rows will be moved to be ordered by date after the user finishes in the date cell. 'sections' being my concept rather than a HOT one - I have users selecting groups of rows and marking them as a 'section'. At the moment the table 'sorts' to the user's eyes, the rows move, and instance.getData() says they're moved, but `data` disagrees which, because of the binding, means the changes to the user seem to 'undo' later - all the moves reverse themselves.

I also went into this on an accidental dupe ticket: https://forum.handsontable.com/t/row-moves-cannot-be-made-permanent/5213 - with a fiddle

I need to find a way to make row moves permanent as soon as they happen, and also preserve the undo/redo buffer - any thoughts would be greatly appreciated

**Comment by henricook** [comment]
@AMBudnik So desperate am I for this for my users I might need to fork this repository and try to retain the undo buffer on `loadData`... Even if HOT don't intend to fix this if an engineer has a moment to give me some thoughts on how I might bodge around it on my own fork that would be great. Obviously I don't really want to maintain my own fork, but the need for move permanency AND a retained undo buffer is high.

**Comment by AMBudnik** [comment]
I will consult my colleague and come back as soon as possible @henricook

**Comment by swistach** [comment]
Hi @henricook,
I discussed your use case with @wojciechczerniak.
I think that forking Handsontable and custom detaching `undo-redo` from `loadData` might make reloading data even more unstable.

As far as I understand what you'd like to achieve - If I were you, I'd add a column, accessible only via API, where I'd store rows order as an integer. That would separate storing source data and their synchronisation to the backend from the presentation layer (sorting/filtering). Although, it might be helpful only if you use only a relatively small amount of data. The worst scenario has to cover moving the last row to the top. In that case, you have to adjust values in the `order` column for all rows.

Please look at this example: https://jsfiddle.net/bvg7c3yw/.

**Comment by henricook** [comment]
Thanks for these thoughts @swistach - very helpful!

**Comment by budnix** [comment]
Hi @henricook,

Apologies for the long silence on this one. I've re-checked the whole thread against the current release (18.0.0). There are three answers here, including one to the question you actually asked that we never came back with.

**1. The original report is by design, and that hasn't changed.**

Moving a row does not reorder the array you pass as `data`. The order is kept in `IndexMapper` and `sourceData` is left untouched, for the reasons @wojciechczerniak gave above. `getData()` reflects the move, `getSourceData()` does not. Still true in 18.0.0, and not something we plan to change.

**2. The question you asked in 2021 has an answer now.**

You wrote:

> Is there a way to replace the data in the table, without all the reinitialisation and destruction that comes with loadData?

Yes — [`updateData()`](https://handsontable.com/docs/react-data-grid/api/core/#updatedata), added in 11.1.0. It replaces the dataset without resetting state, and unlike `loadData()` it does **not** clear the undo/redo stack. So "permanent row moves *and* a working undo buffer" is no longer the either/or you hit. We shipped that method a few months after this thread went quiet and never came back to tell you — sorry.

**3. Guidance we had queued for the next release was wrong.**

While re-checking this we found that a paragraph we'd written for the row-moving guide — telling React users to sync their bound state from `getData()` inside `afterRowMove` — applies the move a second time. Start with `[A,B,C,D]` and drag row 0 to the end: the grid correctly shows `B,C,D,A`, then syncing state turns it into `C,D,A,B`.

That's the same workaround you were using in 2021, and it's now worse than it was then. Back then that path went through `loadData`, which resets the order, so it worked and only cost you undo. Since `updateData`, the row order is kept and gets applied on top of the already-reordered array. The paragraph hasn't been published yet, so nobody has been misled by it, and we're rewriting it before it ships.

**What to do instead**

Pick based on who owns the row order.

*Handsontable owns it.* Seed the order with [`initialState`](https://handsontable.com/docs/react-data-grid/api/options/#initialstate) (16.1.0+) and call `getData()` when you need to save it. Just never feed `getData()` back into the `data` prop — that's the double move above.

*Your app owns it.* This is closer to what you described wanting. Cancel the move and reorder your own state, so React stays the single source of truth:

```js
beforeRowMove(movedRows, finalIndex, dropIndex, movePossible) {
  if (!movePossible) return;
  setData(prev => /* splice movedRows to finalIndex */);
  return false;
}
```

We also found a real bug while testing that second pattern: cancelling a move with `return false` still records an entry on the undo stack, so the next Ctrl+Z does nothing visible. That one is unrelated to React and we've logged it separately.

**Where this stands**

Two internal tasks are now open: rewriting the React section of the row-moving guide to cover both patterns properly before the next release ships, and fixing the cancelled-move undo entry. We'll close this issue once the documentation lands.

Thanks for pushing on this back in 2021 — the thread was detailed enough that re-reading it five years later turned up two real problems.
